app/javascript/pages/Affiliates/Index.tsx

```diff
@@ -334,60 +260,56 @@ const AffiliateRequestsTable = ({
+export default function AffiliatesIndex() {
+  const props = usePage<Props>().props;
   const loggedInUser = useLoggedInUser();
-  const [searchParams, setSearchParams] = useSearchParams();
+  const [isNavigating, setIsNavigating] = React.useState(false);
 
-  const data = cast<PagedAffiliatesData>(useLoaderData());
-  const [affiliateRequests] = React.useState(data.affiliate_requests);
-  const { allow_approve_all_requests: allowApproveAllRequests, affiliates, pagination } = data;
+  const { affiliates, affiliate_requests, pagination, allow_approve_all_requests, affiliates_disabled_reason } = props;
   const [selectedAffiliate, setSelectedAffiliate] = React.useState<Affiliate | null>(null);
   const [sort, setSort] = React.useState<Sort<SortKey> | null>(null);
+
+  const searchParams = new URLSearchParams(window.location.search);
   const searchQuery = searchParams.get("query") ?? "";
 
   const onSearch = useDebouncedCallback((newQuery: string) => {
-    if (searchParams.get("query") === newQuery) return;
-
-    setSearchParams((prevState) => {
-      const params = new URLSearchParams(prevState);
-      if (newQuery.length > 0) {
-        params.set("query", newQuery);
-      } else {
-        params.delete("query");
-      }
-      params.delete("page");
-      return params;
+    const params = new URLSearchParams(window.location.search);
+    if (newQuery.length > 0) {
+      params.set("query", newQuery);
+    } else {
+      params.delete("query");
+    }
+    params.delete("page");
+    setIsNavigating(true);
+    router.visit(`${Routes.affiliates_path()}?${params.toString()}`, {
+      preserveState: true,
+      onFinish: () => setIsNavigating(false),
     });
   }, 500);
 
   const onChangePage = (newPage: number) => {
-    setSearchParams((prevState) => {
-      const params = new URLSearchParams(prevState);
-      params.set("page", newPage.toString());
-      return params;
+    const params = new URLSearchParams(window.location.search);
+    params.set("page", newPage.toString());
+    setIsNavigating(true);
+    router.visit(`${Routes.affiliates_path()}?${params.toString()}`, {
+      preserveState: true,
+      onFinish: () => setIsNavigating(false),
     });
   };
 
   const onSetSort = (newSort: Sort<SortKey> | null) => {
-    setSearchParams((prevState) => {
-      const params = new URLSearchParams(prevState);
-      if (pagination.pages > 1) params.set("page", "1");
-      if (newSort) {
-        params.set("column", newSort.key);
-        params.set("sort", newSort.direction);
-      }
-      return params;
-    });
+    const params = new URLSearchParams(window.location.search);
+    if (pagination.pages > 1) params.set("page", "1");
+    if (newSort) {
+      params.set("column", newSort.key);
+      params.set("sort", newSort.direction);
+    }
     setSort(newSort);
+    setIsNavigating(true);
+    router.visit(`${Routes.affiliates_path()}?${params.toString()}`, {
+      preserveState: true,
```

Comment:
Sicne you are explicitly visiting the same page and preserving the state, it would be simpler to use `reload` and pass `params` to the `data` option so it constructs the query string accordingly. Can you check that the following works?

```suggestion
    router.reload({
      data: params,
```